### PR TITLE
feat(logger): add clear state functionality

### PR DESCRIPTION
**Pull request by saragerion**
## Description of your changes

The Logger utility is commonly initialized in the global scope, outside the handler function.
When you attach persistent log attributes through the `persistentLogAttributes` constructor option or via the `appendKeys`, `addPersistentLogAttributes` methods, this data is attached to the Logger instance.  

Due to the [Lambda Execution Context reuse](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-context.html), this means those persistent log attributes may be reused across invocations.

This PR introduces a new flag that developers can enable to "clear the state" across invocations.
If developers want to make sure that persistent attributes added **inside the handler function** code are not persisted across invocations, they can now set the parameter `clearState` as `true`  in the `injectLambdaContext` middleware or decorator.

### How to verify this change

See [Contributing guide](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/CONTRIBUTING.md) for the docs build. 

carbon (3)

### Related issues, RFCs

[#482](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/482)  

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by dreamorosi — APPROVED** [review]
LGTM!

**Comment by flochaz** [comment]
Could we add an e2e test ?

File: packages/commons/tests/utils/e2eUtils.ts [hunk 1]
```diff
@@ -63,12 +63,15 @@ export const generateUniqueName = (name_prefix: string, uuid: string, runtime: s
 export const invokeFunction = async (functionName: string, times: number = 1, invocationMode: 'PARALLEL' | 'SEQUENTIAL' = 'PARALLEL', payload: FunctionPayload = {}): Promise<InvocationLogs[]> => {
   const invocationLogs: InvocationLogs[] = [];
 
-  const promiseFactory = (): Promise<void> => {
+  const promiseFactory = (index?: number): Promise<void> => {
```
Comment (saragerion) [inline packages/commons/tests/utils/e2eUtils.ts:66]: If you have multiple lambda invocations that are SEQUENTIAL in your e2e scenario, you can now know which invocation is it in the Lambda event payload

**Review by dreamorosi — APPROVED** [review]
LGTM

e2e test are passing, thanks for running them:
https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/2536933581

**Review by flochaz — APPROVED** [review]

**Pull request merged**